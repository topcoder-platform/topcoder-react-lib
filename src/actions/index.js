import authActions from './auth';
import smpActions from './smp';
import statsActions from './stats';
import termsActions from './terms';
import directActions from './direct';
import groupActions from './groups';
import errorActions from './errors';
import challengeActions from './challenge';
import profileActions from './profile';
import memberActions from './members';
import memberTaskActions from './member-tasks';
import reviewOpportunityActions from './reviewOpportunity';
import lookupActions from './lookup';

export const actions = {
  auth: authActions.auth,
  smp: smpActions.smp,
  challenge: challengeActions.challenge,
  terms: termsActions.terms,
  stats: statsActions.stats,
  direct: directActions.direct,
  groups: groupActions.groups,
  errors: errorActions.errors,
  profile: profileActions.profile,
  members: memberActions.members,
  memberTasks: memberTaskActions.memberTasks,
  reviewOpportunity: reviewOpportunityActions.reviewOpportunity,
  lookup: lookupActions.lookup,
};

export default undefined;
