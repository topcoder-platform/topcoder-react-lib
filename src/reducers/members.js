/**
 * @module "reducers.members"
 * @desc Reducer for the Redux store segment that holds members data.
 * @todo Document state segment structure.
 */

import _ from 'lodash';
import { handleActions } from 'redux-actions';
import logger from '../utils/logger';
import actions from '../actions/members';
import { fireErrorMessage } from '../utils/errors';

/**
 * Drops information about a member.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onDrop(state, { payload }) {
  if (!state[payload]) return state;
  return _.omit(state, payload);
}

/**
 * Drops all loaded information on members.
 * @return {Object} New state.
 */
function onDropAll() {
  return {};
}

/**
 * Inits the loading of member achievements.
 * @param {Object} state
 * @param {String} action.handle
 * @param {String} action.uuid
 * @return {Object} New state.
 */
function onGetAchievementsInit(state, action) {
  const { handle, uuid } = action.payload;
  let res = state[handle];
  res = res ? _.clone(res) : {};
  res.achievements = { loadingUuid: uuid };
  return { ...state, [handle]: res };
}

/**
 * Finalizes the loading of member achievements.
 * @param {Object} state
 * @param {Object} error
 * @param {Array} payload.data
 * @param {String} payload.handle
 * @param {String} payload.uuid
 * @return {Object} New state.
 */
function onGetAchievementsDone(state, { error, payload }) {
  if (error) {
    logger.error('Failed to load member achievements', payload);
    fireErrorMessage('Failed to load member achievements');
    return state;
  }

  const { data, handle, uuid } = payload;
  if (uuid !== _.get(state[handle], 'achievements.loadingUuid')) return state;
  return {
    ...state,
    [handle]: {
      ...state[handle],
      achievements: { data, timestamp: Date.now() },
    },
  };
}

/**
 * Initializes the loading of member financial information.
 * @param {Object} state
 * @param {String} action.payload.handle
 * @param {String} action.payload.uuid
 * @return {Object} New state.
 */
function onGetFinancesInit(state, action) {
  const { handle, uuid } = action.payload;
  const envelop = {
    ...(state[handle] || {}),
    finances: {
      loadingUuid: uuid,
      timestamp: 0,
    },
  };
  return {
    ...state,
    [handle]: envelop,
  };
}

/**
 * Finalizes a pending loading of member financial information.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetFinancesDone(state, { error, payload }) {
  if (error) {
    logger.error('Failed to get user financial info', payload);
    fireErrorMessage('Failed to get user financial info', '');
    return state;
  }

  const { data, handle, uuid } = payload;
  if (uuid !== _.get(state[handle], 'finances.loadingUuid')) return state;
  return {
    ...state,
    [handle]: {
      ...state[handle],
      finances: {
        data,
        loadingUuid: '',
        timestamp: Date.now(),
      },
    },
  };
}

/**
 * Inits the loading of member stats.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetStatsInit(state, action) {
  const { handle, uuid } = action.payload;
  let res = state[handle];
  res = res ? _.clone(res) : {};
  res.stats = { loadingUuid: uuid };
  return {
    ...state,
    [handle]: res,
  };
}

/**
 * Finalizes the loading of member stats.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetStatsDone(state, { error, payload }) {
  if (error) {
    logger.error('Failed to get member stats', payload);
    fireErrorMessage('Failed to get member stats', '');
    return state;
  }

  const { data, handle, uuid } = payload;
  if (uuid !== _.get(state[handle], 'stats.loadingUuid')) return state;
  return {
    ...state,
    [handle]: {
      ...state[handle],
      stats: { data, timestamp: Date.now() },
    },
  };
}

/**
 * Inits the loading of member stats history.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetStatsHistoryInit(state, action) {
  const { handle, uuid } = action.payload;
  let res = state[handle];
  res = res ? _.clone(res) : {};
  res.statsHistory = { loadingUuid: uuid };
  return {
    ...state,
    [handle]: res,
  };
}

/**
 * Finalizes the loading of member stats history.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetStatsHistoryDone(state, { error, payload }) {
  if (error) {
    logger.error('Failed to get member statsHistory', payload);
    fireErrorMessage('Failed to get member statsHistory', '');
    return state;
  }

  const { data, handle, uuid } = payload;
  if (uuid !== _.get(state[handle], 'statsHistory.loadingUuid')) return state;
  return {
    ...state,
    [handle]: {
      ...state[handle],
      statsHistory: { data, timestamp: Date.now() },
    },
  };
}

/**
 * Inits the loading of member stats distribution.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetStatsDistributionInit(state, action) {
  const { handle, uuid } = action.payload;
  let res = state[handle];
  res = res ? _.clone(res) : {};
  res.statsDistribution = { loadingUuid: uuid };
  return {
    ...state,
    [handle]: res,
  };
}

/**
 * Finalizes the loading of member stats distribution.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetStatsDistributionDone(state, { error, payload }) {
  if (error) {
    logger.error('Failed to get member statsDistribution', payload);
    fireErrorMessage('Failed to get member statsDistribution', '');
    return state;
  }

  const { data, handle, uuid } = payload;
  if (uuid !== _.get(state[handle], 'statsDistribution.loadingUuid')) return state;
  return {
    ...state,
    [handle]: {
      ...state[handle],
      statsDistribution: { data, timestamp: Date.now() },
    },
  };
}

/**
 * Inits the loading of member active challenges.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetActiveChallengesInit(state, action) {
  const { handle } = action.payload;
  return {
    ...state,
    [handle]: { ...state[handle], activeChallengesCount: null },
  };
}

/**
 * Finalizes the loading of member active challenges.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetActiveChallengesDone(state, { error, payload }) {
  if (error) {
    logger.error('Failed to get member active challenges', payload);
    fireErrorMessage('Failed to get member active challenges', '');
    return state;
  }

  const { handle, challenges } = payload;

  return {
    ...state,
    [handle]: {
      ...state[handle],
      activeChallengesCount: challenges.length,
    },
  };
}

/**
 * Inits the loading of member subtrack challenges.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetSubtrackChallengesInit(state, { payload }) {
  const { handle, uuid } = payload;
  return {
    ...state,
    [handle]: { ...state[handle], loadingSubTrackChallengesUUID: uuid },
  };
}

/**
 * Finalizes the loading of member subtrack challenges.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetSubtrackChallengesDone(state, { error, payload }) {
  if (error) {
    logger.error('Failed to get member subtrack challenges', payload);
    fireErrorMessage('Failed to get member subtrack challenges', '');
    return state;
  }

  const {
    uuid,
    challenges,
    refresh,
    handle,
  } = payload;
  if (uuid !== state[handle].loadingSubTrackChallengesUUID) return state;

  return {
    ...state,
    [handle]: {
      ...state[handle],
      subtrackChallenges: (state[handle].subtrackChallenges && !refresh)
        ? [...state[handle].subtrackChallenges, ...challenges]
        : challenges,
      // if current query returns 0 item, mark it completed
      subtrackChallengesHasMore: challenges && challenges.length > 0,
      loadingSubTrackChallengesUUID: '',
    },
  };
}

/**
 * Inits the loading of member SRMs.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetUserSRMInit(state, { payload }) {
  const { handle, uuid } = payload;
  return {
    ...state,
    [handle]: { ...state[handle], loadingSRMUUID: uuid },
  };
}

/**
 * Finalizes the loading of member SRMs.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetUserSRMDone(state, { error, payload }) {
  if (error) {
    logger.error('Failed to get member SRMs', payload);
    fireErrorMessage('Failed to get member SRMs', '');
    return state;
  }

  const {
    uuid,
    srms,
    refresh,
    handle,
  } = payload;
  if (uuid !== state[handle].loadingSRMUUID) return state;

  return {
    ...state,
    [handle]: {
      ...state[handle],
      userSRMs: (state[handle].userSRMs && !refresh) ? [...state[handle].userSRMs, ...srms] : srms,
      userSRMHasMore: srms && srms.length > 0, // if current query returns 0 item, mark it completed
      loadingSRMUUID: '',
    },
  };
}

/**
 * Inits the loading of member marathons.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetUserMarathonInit(state, { payload }) {
  const { handle, uuid } = payload;
  return {
    ...state,
    [handle]: { ...state[handle], loadingMarathonUUID: uuid },
  };
}

/**
 * Finalizes the loading of member marathons.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetUserMarathonDone(state, { error, payload }) {
  if (error) {
    logger.error('Failed to get member marathons', payload);
    fireErrorMessage('Failed to get member marathons', '');
    return state;
  }

  const {
    uuid,
    marathons,
    refresh,
    handle,
  } = payload;
  if (uuid !== state[handle].loadingMarathonUUID) return state;

  return {
    ...state,
    [handle]: {
      ...state[handle],
      userMarathons: (state[handle].userMarathons && !refresh)
        ? [...state[handle].userMarathons, ...marathons.challenges]
        : marathons.challenges,
      // if current query returns 0 item, mark it completed
      userMarathonHasMore: marathons && marathons.challenges && marathons.challenges.length > 0,
      loadingMarathonUUID: '',
    },
  };
}

/**
 * Inits the loading of user challenge resources.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetUserResourcesInit(state, { payload }) {
  const { uuid } = payload;
  return {
    ...state,
    userResources: { resources: [], loadingUserResources: uuid },
  };
}

/**
 * Finalizes the loading of user challenge resources.
 * @param {Object} state
 * @param {Object} action
 * @return {Object} New state.
 */
function onGetUserResourcesDone(state, { error, payload }) {
  if (error) {
    logger.error('Failed to get user resources', payload);
    fireErrorMessage('Failed to get user resources', '');
    return state;
  }

  const { uuid, resources } = payload;

  if (uuid !== state.userResources.loadingUserResources) return state;

  return {
    ...state,
    userResources: { resources, loadingUserResources: '' },
  };
}

/**
 * Creates a new Members reducer with the specified initial state.
 * @param {Object} initialState Optional. Initial state.
 * @return {Function} Members reducer.
 */
function create(initialState = {}) {
  const a = actions.members;
  return handleActions({
    [a.drop]: onDrop,
    [a.dropAll]: onDropAll,
    [a.getAchievementsInit]: onGetAchievementsInit,
    [a.getAchievementsDone]: onGetAchievementsDone,
    [a.getAchievementsV3Done]: onGetAchievementsDone,
    [a.getFinancesInit]: onGetFinancesInit,
    [a.getFinancesDone]: onGetFinancesDone,
    [a.getStatsInit]: onGetStatsInit,
    [a.getStatsDone]: onGetStatsDone,
    [a.getStatsHistoryInit]: onGetStatsHistoryInit,
    [a.getStatsHistoryDone]: onGetStatsHistoryDone,
    [a.getStatsDistributionInit]: onGetStatsDistributionInit,
    [a.getStatsDistributionDone]: onGetStatsDistributionDone,
    [a.getActiveChallengesInit]: onGetActiveChallengesInit,
    [a.getActiveChallengesDone]: onGetActiveChallengesDone,
    [a.getSubtrackChallengesInit]: onGetSubtrackChallengesInit,
    [a.getSubtrackChallengesDone]: onGetSubtrackChallengesDone,
    [a.getUserSrmInit]: onGetUserSRMInit,
    [a.getUserSrmDone]: onGetUserSRMDone,
    [a.getUserMarathonInit]: onGetUserMarathonInit,
    [a.getUserMarathonDone]: onGetUserMarathonDone,
    [a.getUserResourcesInit]: onGetUserResourcesInit,
    [a.getUserResourcesDone]: onGetUserResourcesDone,
  }, initialState);
}

/**
 * Factory which creates a new reducer with its initial state tailored to the
 * given options object, if specified (for server-side rendering). If options
 * object is not specified, it creates just the default reducer. Accepted options are:
 * @return {Promise}
 * @resolves {Function(state, action): state} New reducer.
 */
export function factory() {
  return Promise.resolve(create());
}

/**
 * @static
 * @member default
 * @desc Reducer with default initial state.
 */
export default create();
