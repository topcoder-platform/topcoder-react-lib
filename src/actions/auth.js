/**
 * @module "actions.auth"
 * @desc Actions related to Topcoder authentication system.
 */

import { createActions } from 'redux-actions';
import { decodeToken } from 'tc-accounts';
import { getApi } from '../services/api';

/**
 * @static
 * @desc Creates an action that loads Topcoder user profile from v3 API.
 * @param {String} userTokenV3 v3 authentication token.
 * @return {Action}
 */
function loadProfileDone(userTokenV3) {
  if (!userTokenV3) return Promise.resolve(null);
  const user = decodeToken(userTokenV3);
  const api = getApi('V3', userTokenV3);
  const apiV5 = getApi('V5', userTokenV3);
  return Promise.all([
    api.get(`/members/${user.handle}`)
      .then(res => res.json()).then(res => (res.result.status === 200 ? res.result.content : {})),
    apiV5.get(`/groups?memberId=${user.userId}&membershipType=user`)
      .then(res => (res.ok ? res.json() : new Error(res.statusText))),
  ]).then(([profile, groups]) => ({ ...profile, groups }));
}

/**
 * @static
 * @desc Creates an action that sets Topcoder v2 authentication token.
 * @param {String} tokenV2 Topcoder v2 authentication token.
 * @return {Action}
 */
function setTcTokenV2(tokenV2) {
  return tokenV2;
}

/**
 * @static
 * @desc Creates an action that decodes Topcoder v3 authentication token,
 * to get user object, and then writes both the token and the user object into
 * Redux store.
 * @param {String} tokenV3 Topcoder v3 authentication token.
 * @return {Action}
 */
function setTcTokenV3(tokenV3) {
  return tokenV3;
}

export default createActions({
  AUTH: {
    LOAD_PROFILE: loadProfileDone,
    SET_TC_TOKEN_V2: setTcTokenV2,
    SET_TC_TOKEN_V3: setTcTokenV3,
  },
});
