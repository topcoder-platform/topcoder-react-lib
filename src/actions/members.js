/**
 * @module "actions.members"
 * @desc Actions related to members data.
 */

import { createActions } from 'redux-actions';
import { getService } from '../services/members';
import { getService as getUserService } from '../services/user';
import { getService as getChallengesService } from '../services/challenges';

/**
 * @static
 * @desc Creates an action that drops all information related to the specfied
 *  member.
 * @param {String} handle Topcoder user handle.
 * @return {Action}
 */
function drop(handle) {
  return handle;
}

/**
 * @static
 * @desc Creates an action that drops all member information loaded by
 *  actions from this module.
 * @return {Action}
 */
function dropAll() {}

/**
 * @static
 * @desc Creates an action that signals beginning of member achievements
 *  loading.
 * @param {String} handle Member handle.
 * @param {String} uuid Loading operation UUID.
 * @return {Action}
 */
function getAchievementsInit(handle, uuid) {
  return { handle, uuid };
}

/**
 * @static
 * @desc Creates an action that loads member achievements.
 * @param {String} handle Member handle.
 * @param {String} uuid Loading operation
 * @return {Action}
 */
async function getAchievementsDone(handle, uuid) {
  let data;
  try {
    data = await getUserService().getAchievements(handle);
  } catch (e) {
    data = [];
  }
  return { data, handle, uuid };
}

/**
 * @static
 * @desc Creates an action that signals beginning of loading the member's
 *  financial information.
 * @param {String} handle Member handle.
 * @param {String} uuid Operation UUID.
 * @return {Action}
 */
function getFinancesInit(handle, uuid) {
  return { handle, uuid };
}

/**
 * @static
 * @desc Creates an action that loads member's financial information.
 * @param {String} handle Member handle.
 * @param {String} uuid Operation UUID.
 * @param {String} tokenV3 v3 auth token.
 * @return {Action}
 */
async function getFinancesDone(handle, uuid, tokenV3) {
  const data = await getService(tokenV3).getMemberFinances(handle);
  return { data, handle, uuid };
}

/**
 * @static
 * @desc Creates an action that signals beginning of member stats loading.
 * @param {String} handle Member handle.
 * @param {String} uuid Operation UUID.
 * @return {Action}
 */
async function getStatsInit(handle, uuid) {
  return { handle, uuid };
}

/**
 * @static
 * @desc Create an action that loads member statistics.
 * @param {String} handle Member handle.
 * @param {String} uuid Operation UUID.
 * @param {String} tokenV3 v3 auth token.
 * @return {Action}
 */
async function getStatsDone(handle, uuid, tokenV3) {
  const data = await getService(tokenV3).getStats(handle);
  return { data, handle, uuid };
}

/**
 * @static
 * @desc Payload creator for the action that inits the loading of member active challenges.
 * @param {String} handle
 * @param {String} uuid
 * @returns {Object} Payload
 */
async function getActiveChallengesInit(handle, uuid) {
  return { handle, uuid };
}

/**
 * @static
 * @desc Payload creator for the action that loads the member active challenges.
 * @param {String} handle
 * @param {String} uuid
 * @param {String} tokenV3
 * @returns {Object} Payload
 */
async function getActiveChallengesDone(handle, uuid, tokenV3) {
  const filter = { status: 'ACTIVE' };
  const service = getChallengesService(tokenV3);
  /* TODO: Reuse `getAll` from `actions/challenge-listing`
  /* after it moved from `community-app` to here.
   */
  function getAll(getter, page = 0, prev = null) {
    const PAGE_SIZE = 50;
    return getter({
      limit: PAGE_SIZE,
      offset: page * PAGE_SIZE,
    }).then(({ challenges: chunk }) => {
      if (!chunk.length) return prev || [];
      return getAll(getter, 1 + page, prev ? prev.concat(chunk) : chunk);
    });
  }
  const calls = [
    getAll(params =>
      service.getUserChallenges(handle, filter, params)),
    getAll(params =>
      service.getUserMarathonMatches(handle, filter, params)),
  ];

  const [challenges] = await Promise.all(calls);

  return { handle, challenges, uuid };
}

/**
 * @static
 * @desc Create an action that signals beginning of member stats distribution history.
 * @param {String} handle Member handle.
 * @param {String} uuid Operation UUID.
 * @return {Action}
 */
async function getStatsHistoryInit(handle, uuid) {
  return { handle, uuid };
}

/**
 * @static
 * @desc Create an action that loads the member stats history.
 * @param {String} handle Member handle.
 * @param {String} uuid Operation UUID.
 * @param {String} tokenV3 v3 auth token.
 * @return {Action}
 */
async function getStatsHistoryDone(handle, uuid, tokenV3) {
  const data = await getService(tokenV3).getStatsHistory(handle);
  return { data, handle, uuid };
}

/**
 * @static
 * @desc Create an action that signals beginning of member stats distribution loading.
 * @param {String} handle Member handle.
 * @param {String} uuid Operation UUID.
 * @return {Action}
 */
async function getStatsDistributionInit(handle, uuid) {
  return { handle, uuid };
}

/**
 * @static
 * @desc Create an action that loads the member stats distribution.
 * @param {String} handle Member handle.
 * @param {String} track Main track name.
 * @param {String} subTrack Subtrack name.
 * @param {String} uuid Operation UUID.
 * @param {String} tokenV3 v3 auth token.
 * @return {Action}
 */
async function getStatsDistributionDone(handle, track, subTrack, uuid, tokenV3) {
  const data = await getService(tokenV3).getStatsDistribution(handle, track, subTrack);
  return { data, handle, uuid };
}

export default createActions({
  MEMBERS: {
    DROP: drop,
    DROP_ALL: dropAll,
    GET_ACHIEVEMENTS_INIT: getAchievementsInit,
    GET_ACHIEVEMENTS_DONE: getAchievementsDone,
    GET_FINANCES_INIT: getFinancesInit,
    GET_FINANCES_DONE: getFinancesDone,
    GET_STATS_INIT: getStatsInit,
    GET_STATS_DONE: getStatsDone,
    GET_STATS_HISTORY_INIT: getStatsHistoryInit,
    GET_STATS_HISTORY_DONE: getStatsHistoryDone,
    GET_STATS_DISTRIBUTION_INIT: getStatsDistributionInit,
    GET_STATS_DISTRIBUTION_DONE: getStatsDistributionDone,
    GET_ACTIVE_CHALLENGES_INIT: getActiveChallengesInit,
    GET_ACTIVE_CHALLENGES_DONE: getActiveChallengesDone,
  },
});
