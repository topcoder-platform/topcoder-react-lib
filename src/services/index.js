/**
 * Export services.
 */
import * as api from './api';
import * as billing from './billing';
import * as challenge from './challenges';
import * as direct from './direct';
import * as groups from './groups';
import * as members from './members';
import * as terms from './terms';
import * as communities from './communities';
import * as reviewOpportunities from './reviewOpportunities';
import * as userSetting from './user-settings';
import * as user from './user';
import * as lookup from './lookup';

export const services = {
  api,
  billing,
  terms,
  communities,
  challenge,
  direct,
  groups,
  members,
  user,
  userSetting,
  reviewOpportunities,
  lookup,
};

export default undefined;
