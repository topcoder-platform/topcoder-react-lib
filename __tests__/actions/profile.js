import { redux } from 'topcoder-react-utils';

import * as MembersService from 'services/members';
import * as UserService from 'services/user';

import actions from 'actions/profile';

const handle = 'tcscoder';
const tokenV3 = 'tokenV3';
const profile = { userId: 12345, handle };
const skill = { tagId: 123, tagName: 'Node.js' };
const weblink = 'https://www.google.com';
const linkedAccounts = [{
  providerType: 'github',
  social: true,
  userId: '623633',
}];

// Mock services
const mockMembersService = {
  getPresignedUrl: jest.fn().mockReturnValue(Promise.resolve()),
  uploadFileToS3: jest.fn().mockReturnValue(Promise.resolve()),
  updateMemberPhoto: jest.fn().mockReturnValue(Promise.resolve('url-of-photo')),
  updateMemberProfile: jest.fn().mockReturnValue(Promise.resolve(profile)),
  addSkill: jest.fn().mockReturnValue(Promise.resolve({ skills: [skill] })),
  hideSkill: jest.fn().mockReturnValue(Promise.resolve({ skills: [] })),
  addWebLink: jest.fn().mockReturnValue(Promise.resolve(weblink)),
  deleteWebLink: jest.fn().mockReturnValue(Promise.resolve(weblink)),
};
MembersService.getService = jest.fn().mockReturnValue(mockMembersService);

const mockUserService = {
  linkExternalAccount: jest.fn().mockReturnValue(Promise.resolve(linkedAccounts[0])),
  unlinkExternalAccount: jest.fn().mockReturnValue(Promise.resolve('unlinked')),
  getLinkedAccounts: jest.fn().mockReturnValue(Promise.resolve({ profiles: linkedAccounts })),
  getCredential: jest.fn().mockReturnValue(Promise.resolve({ credential: { hasPassword: true } })),
  getEmailPreferences:
    jest.fn().mockReturnValue(Promise.resolve({ subscriptions: { TOPCODER_NL_DATA: true } })),
  saveEmailPreferences:
    jest.fn().mockReturnValue(Promise.resolve({ subscriptions: { TOPCODER_NL_DATA: true } })),
  updatePassword: jest.fn().mockReturnValue(Promise.resolve({ update: true })),
};
UserService.getService = jest.fn().mockReturnValue(mockUserService);

test('Module exports', () => expect(actions).toMatchSnapshot());

test('profile.uploadPhotoDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.profile.uploadPhotoDone(handle, tokenV3));
  expect(actionResult).toMatchSnapshot();
  expect(mockMembersService.getPresignedUrl).toBeCalled();
  expect(mockMembersService.uploadFileToS3).toBeCalled();
  expect(mockMembersService.updateMemberPhoto).toBeCalled();
});

test('profile.updateProfileDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.profile.updateProfileDone(profile, tokenV3));
  expect(actionResult).toMatchSnapshot();
  expect(mockMembersService.updateMemberProfile).toBeCalled();
});

test('profile.addSkillDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.profile.addSkillDone(handle, tokenV3, skill));
  expect(actionResult).toMatchSnapshot();
  expect(mockMembersService.addSkill).toBeCalled();
});

test('profile.hideSkillDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.profile.hideSkillDone(handle, tokenV3, skill));
  expect(actionResult).toMatchSnapshot();
  expect(mockMembersService.hideSkill).toBeCalled();
});

test('profile.addWebLinkDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.profile.addWebLinkDone(handle, tokenV3, weblink));
  expect(actionResult).toMatchSnapshot();
  expect(mockMembersService.addWebLink).toBeCalled();
});

test('profile.deleteWebLinkDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.profile.deleteWebLinkDone(handle, tokenV3, weblink));
  expect(actionResult).toMatchSnapshot();
  expect(mockMembersService.deleteWebLink).toBeCalled();
});

test('profile.linkExternalAccountDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.profile.linkExternalAccountDone(profile, tokenV3, 'github'));
  expect(actionResult).toMatchSnapshot();
  expect(mockUserService.linkExternalAccount).toBeCalled();
});

test('profile.unlinkExternalAccountDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.profile.unlinkExternalAccountDone(profile, tokenV3, 'github'));
  expect(actionResult).toMatchSnapshot();
  expect(mockUserService.unlinkExternalAccount).toBeCalled();
});

test('profile.getLinkedAccountsDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.profile.getLinkedAccountsDone(profile, tokenV3));
  expect(actionResult).toMatchSnapshot();
  expect(mockUserService.getLinkedAccounts).toBeCalled();
});

test('profile.getCredentialDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.profile.getCredentialDone(profile, tokenV3));
  expect(actionResult).toMatchSnapshot();
  expect(mockUserService.getCredential).toBeCalled();
});

test('profile.getEmailPreferencesDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.profile.getEmailPreferencesDone(profile, tokenV3));
  expect(actionResult).toMatchSnapshot();
  expect(mockUserService.getEmailPreferences).toBeCalled();
});

test('profile.saveEmailPreferencesDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.profile.saveEmailPreferencesDone(profile, tokenV3, {}));
  expect(actionResult).toMatchSnapshot();
  expect(mockUserService.saveEmailPreferences).toBeCalled();
});

test('profile.updatePasswordDone', async () => {
  const actionResult =
    await redux.resolveAction(actions.profile.updatePasswordDone(profile, tokenV3, 'newPassword', 'oldPassword'));
  expect(actionResult).toMatchSnapshot();
  expect(mockUserService.updatePassword).toBeCalled();
});

