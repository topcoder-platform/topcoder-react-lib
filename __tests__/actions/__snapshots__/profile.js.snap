// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module exports 1`] = `
Object {
  "profile": Object {
    "addSkillDone": [Function],
    "addSkillInit": [Function],
    "addWebLinkDone": [Function],
    "addWebLinkInit": [Function],
    "deletePhotoDone": [Function],
    "deletePhotoInit": [Function],
    "deleteWebLinkDone": [Function],
    "deleteWebLinkInit": [Function],
    "getAchievementsDone": [Function],
    "getAchievementsInit": [Function],
    "getCredentialDone": [Function],
    "getCredentialInit": [Function],
    "getEmailPreferencesDone": [Function],
    "getEmailPreferencesInit": [Function],
    "getExternalAccountsDone": [Function],
    "getExternalAccountsInit": [Function],
    "getExternalLinksDone": [Function],
    "getExternalLinksInit": [Function],
    "getInfoDone": [Function],
    "getInfoInit": [Function],
    "getLinkedAccountsDone": [Function],
    "getLinkedAccountsInit": [Function],
    "getSkillsDone": [Function],
    "getSkillsInit": [Function],
    "getStatsDone": [Function],
    "getStatsInit": [Function],
    "hideSkillDone": [Function],
    "hideSkillInit": [Function],
    "linkExternalAccountDone": [Function],
    "linkExternalAccountInit": [Function],
    "loadProfile": [Function],
    "saveEmailPreferencesDone": [Function],
    "saveEmailPreferencesInit": [Function],
    "unlinkExternalAccountDone": [Function],
    "unlinkExternalAccountInit": [Function],
    "updatePasswordDone": [Function],
    "updatePasswordInit": [Function],
    "updateProfileDone": [Function],
    "updateProfileInit": [Function],
    "uploadPhotoDone": [Function],
    "uploadPhotoInit": [Function],
  },
}
`;

exports[`profile.addSkillDone 1`] = `
Object {
  "payload": Object {
    "handle": "tcscoder",
    "skill": Object {
      "tagId": 123,
      "tagName": "Node.js",
    },
    "skills": Array [
      Object {
        "tagId": 123,
        "tagName": "Node.js",
      },
    ],
  },
  "type": "PROFILE/ADD_SKILL_DONE",
}
`;

exports[`profile.addWebLinkDone 1`] = `
Object {
  "payload": Object {
    "data": "https://www.google.com",
    "handle": "tcscoder",
  },
  "type": "PROFILE/ADD_WEB_LINK_DONE",
}
`;

exports[`profile.deleteWebLinkDone 1`] = `
Object {
  "payload": Object {
    "data": "https://www.google.com",
    "handle": "tcscoder",
  },
  "type": "PROFILE/DELETE_WEB_LINK_DONE",
}
`;

exports[`profile.getCredentialDone 1`] = `
Object {
  "payload": Object {
    "credential": Object {
      "hasPassword": true,
    },
  },
  "type": "PROFILE/GET_CREDENTIAL_DONE",
}
`;

exports[`profile.getEmailPreferencesDone 1`] = `
Object {
  "payload": Object {
    "subscriptions": Object {
      "TOPCODER_NL_DATA": true,
    },
  },
  "type": "PROFILE/GET_EMAIL_PREFERENCES_DONE",
}
`;

exports[`profile.getLinkedAccountsDone 1`] = `
Object {
  "payload": Object {
    "profiles": Array [
      Object {
        "providerType": "github",
        "social": true,
        "userId": "623633",
      },
    ],
  },
  "type": "PROFILE/GET_LINKED_ACCOUNTS_DONE",
}
`;

exports[`profile.hideSkillDone 1`] = `
Object {
  "payload": Object {
    "handle": "tcscoder",
    "skill": Object {
      "tagId": 123,
      "tagName": "Node.js",
    },
    "skills": Array [],
  },
  "type": "PROFILE/HIDE_SKILL_DONE",
}
`;

exports[`profile.linkExternalAccountDone 1`] = `
Object {
  "payload": Object {
    "data": Object {
      "providerType": "github",
      "social": true,
      "userId": "623633",
    },
    "handle": "tcscoder",
  },
  "type": "PROFILE/LINK_EXTERNAL_ACCOUNT_DONE",
}
`;

exports[`profile.saveEmailPreferencesDone 1`] = `
Object {
  "payload": Object {
    "data": Object {
      "subscriptions": Object {
        "TOPCODER_NL_DATA": true,
      },
    },
    "handle": "tcscoder",
  },
  "type": "PROFILE/SAVE_EMAIL_PREFERENCES_DONE",
}
`;

exports[`profile.unlinkExternalAccountDone 1`] = `
Object {
  "payload": Object {
    "handle": "tcscoder",
    "providerType": "github",
  },
  "type": "PROFILE/UNLINK_EXTERNAL_ACCOUNT_DONE",
}
`;

exports[`profile.updatePasswordDone 1`] = `
Object {
  "payload": Object {
    "data": Object {
      "update": true,
    },
    "handle": "tcscoder",
  },
  "type": "PROFILE/UPDATE_PASSWORD_DONE",
}
`;

exports[`profile.updateProfileDone 1`] = `
Object {
  "payload": Object {
    "handle": "tcscoder",
    "userId": 12345,
  },
  "type": "PROFILE/UPDATE_PROFILE_DONE",
}
`;

exports[`profile.uploadPhotoDone 1`] = `
Object {
  "payload": Object {
    "handle": "tcscoder",
    "photoURL": "url-of-photo",
  },
  "type": "PROFILE/UPLOAD_PHOTO_DONE",
}
`;
